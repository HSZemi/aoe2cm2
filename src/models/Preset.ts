import Civilisation from "./Civilisation";
import Turn from "./Turn";
import {CivilisationEncoder} from "../util/CivilisationEncoder";
import {Assert} from "../util/Assert";

class Preset {

    public static readonly EMPTY: Preset = new Preset('', [], []);

    public static readonly NEW: Preset = new Preset('Change me', Civilisation.ALL, []);

    public static readonly SAMPLE: Preset = new Preset('Default Preset', Civilisation.ALL, [
        Turn.HOST_GLOBAL_BAN,
        Turn.GUEST_GLOBAL_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_GLOBAL_PICK,
        Turn.GUEST_GLOBAL_PICK,
        Turn.HOST_PICK,
        Turn.GUEST_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_SNIPE,
        Turn.GUEST_HIDDEN_SNIPE,
        Turn.REVEAL_ALL
    ]);

    public static readonly SIMPLE: Preset = new Preset('Simple Preset', Civilisation.ALL, [
        Turn.HOST_NONEXCLUSIVE_BAN,
        Turn.GUEST_NONEXCLUSIVE_BAN,
        Turn.GUEST_NONEXCLUSIVE_PICK,
        Turn.HOST_NONEXCLUSIVE_PICK
    ]);
    public readonly name: string;
    public readonly encodedCivilisations: string;
    public readonly turns: Turn[];

    constructor(name: string, civilisations: Civilisation[], turns: Turn[] = []) {
        this.name = name;
        this.encodedCivilisations = CivilisationEncoder.encodeCivilisationArray(civilisations);
        this.turns = turns;
    }

    public static fromPojo(preset: { name: string, encodedCivilisations: string, turns: Turn[] } | undefined): Preset | undefined {
        if (preset === undefined) {
            return undefined;
        }
        Assert.isString(preset.name);
        Assert.isString(preset.encodedCivilisations);
        return new Preset(preset.name, CivilisationEncoder.decodeCivilisationArray(preset.encodedCivilisations), Turn.fromPojoArray(preset.turns));
    }

    public addTurn(turn: Turn) {
        this.turns.push(turn);
    }

    get civilisations(): Civilisation[] {
        return CivilisationEncoder.decodeCivilisationArray(this.encodedCivilisations);
    }
}

export default Preset;