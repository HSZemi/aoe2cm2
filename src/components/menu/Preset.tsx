import * as React from "react";
import {PresetUtil} from "../../util/PresetUtil";
import {default as ModelPreset} from "../../models/Preset"
import TurnRow from "../draft/TurnRow";
import NewDraftButton from "../NewDraftButton";
import CustomisePresetButton from "../../containers/CustomisePresetButton";
import Civilisation from "../../models/Civilisation";
import CopyableInput from "../draft/CopyableInput";

interface IState {
    preset?: ModelPreset;
    presetExists: boolean;
}

class Preset extends React.Component<object, IState> {

    state: IState = {presetExists: true};

    componentDidMount() {
        const presetId = PresetUtil.getIdFromUrl();
        if (presetId !== undefined) {
            this.loadPreset(presetId);
        }
    }

    public render() {
        if (!this.state.presetExists) {
            return (
                <div>
                    <p>Could not find this preset, sorry :-(</p>
                </div>
            );
        }
        if (this.state.preset !== undefined) {

            const presetCivilisations = this.state.preset.civilisations;
            const civs = Civilisation.ALL.map((value: Civilisation, index: number) => {
                return <div className="pure-u-1-4">
                    <label>
                        <input type='checkbox' checked={presetCivilisations.includes(value)} disabled/> {value.name}
                    </label>
                </div>;
            });

            return (
                <div className='box centered'>
                    <h3>{this.state.preset.name}</h3>

                    <TurnRow turns={this.state.preset.turns}/>

                    <div className="pure-g">
                        {civs}
                    </div>
                    <p>
                        <NewDraftButton preset={this.state.preset}/>
                    </p>
                    <p className={'wide-input'}>
                        <CopyableInput content={window.location.href} before={'preset.shareThisPreset'}/>
                    </p>
                    <p>
                        <CustomisePresetButton preset={this.state.preset}/>
                    </p>
                </div>
            );
        }
        return (
            <div>
                <p>Loadingâ€¦</p>
            </div>
        );
    }


    private loadPreset = (presetId: string) => {
        const request = new XMLHttpRequest();
        request.open('GET', '/getpreset/' + presetId, true);
        request.onreadystatechange = () => {
            if (request.readyState === XMLHttpRequest.DONE && request.status === 200) {
                const result = JSON.parse(request.responseText);
                this.setState({preset: ModelPreset.fromPojo(result)});
            } else if (request.readyState === XMLHttpRequest.DONE && request.status === 404) {
                this.setState({presetExists: false});
            }
        };
        request.send();
    };
}

export default Preset;