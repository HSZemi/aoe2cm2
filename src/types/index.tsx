import Preset from "../models/Preset";
import Player from "../models/Player";
import {DraftEvent} from "../models/DraftEvent";
export interface IStoreState {
    nameHost?: string;
    nameGuest?: string;
    hostReady: boolean;
    guestReady: boolean;
    whoAmI?: Player;
    preset?: Preset;
    nextAction: number;
    events: DraftEvent[];
}
