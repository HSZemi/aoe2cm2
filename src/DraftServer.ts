import express from "express";
import {Server} from "http"
import socketio from "socket.io";
import Player from "./constants/Player";
import {IDraftConfig} from "./types/IDraftConfig";
import {ISetRoleMessage} from "./types/ISetRoleMessage";
import {DraftsStore} from "./models/DraftsStore";
import {Validator} from "./models/Validator";
import {ValidationId} from "./constants/ValidationId";
import {DraftEvent} from "./types/DraftEvent";
import {Util} from "./util/Util";
import {AddressInfo} from "net";
import Preset from "./models/Preset";
import {Listeners} from "./util/Listeners";
import * as fs from "fs";
import {logger} from "./util/Logger";
import {ISetNameMessage} from "./types/ISetNameMessage";
import {PresetUtil} from "./util/PresetUtil";

const TWELVE_HOURS = 1000 * 60 * 60 * 12;

function getAssignedRole(socket: SocketIO.Socket, roomHost: string, roomGuest: string): Player {
    let assignedRole: Player = Player.NONE;
    if (Object.keys(socket.rooms).includes(roomHost)) {
        assignedRole = Player.HOST;
    } else if (Object.keys(socket.rooms).includes(roomGuest)) {
        assignedRole = Player.GUEST;
    }
    return assignedRole;
}

export const DraftServer = {
    serve(port: string | number | undefined): { httpServerAddr: AddressInfo | string | null; io: SocketIO.Server; httpServer: Server } {
        logger.info("Starting DraftServer on port %d", port);
        const app = express();
        app.set("port", port);
        app.use(express.json());

        const server = new Server(app);
        const io = socketio(server, {cookie: false});
        const draftsStore = new DraftsStore();
        const validator = new Validator(draftsStore);

        function connectPlayer(draftId: string, player: Player, name: string) {
            draftsStore.connectPlayer(draftId, player, name);
        }

        function setPlayerName(draftId: string, player: Player, name: string) {
            draftsStore.setPlayerName(draftId, player, name);
        }

        app.post('/draft/new', (req, res) => {
            logger.info('Received request to create a new draft: %s', JSON.stringify(req.body));
            let draftId = Util.newDraftId();
            while (draftsStore.has(draftId)) {
                draftId += Util.randomChar();
            }
            const pojo: Preset = req.body.preset as Preset;
            let preset = Preset.fromPojo(pojo);
            const validationErrors = Validator.validatePreset(preset);
            if (validationErrors.length === 0) {
                draftsStore.initDraft(draftId, preset as Preset);
                res.json({status: 'ok', draftId});
                logger.info('Created new draft with id: %s', draftId, {draftId});
            } else {
                res.json({status: 'error', validationErrors});
                logger.info('Draft validation failed: %s', JSON.stringify(validationErrors));
            }
        });
        app.post('/preset/new', (req, res) => {
            logger.info('Received request to create a new preset: %s', JSON.stringify(req.body));
            const pojo: Preset = req.body.preset as Preset;
            let preset = Preset.fromPojo(pojo);
            const validationErrors = Validator.validatePreset(preset);
            if (validationErrors.length === 0) {
                const presetId = PresetUtil.createPreset(preset as Preset);
                res.json({status: 'ok', presetId: presetId});
                logger.info('Created new preset with id: %s', presetId, {presetId});
            } else {
                res.json({status: 'error', validationErrors});
                logger.info('Preset validation failed: %s', JSON.stringify(validationErrors));
            }
        });
        app.get('/getpreset/list', (req, res) => {
            res.sendFile('presets.json', {'root': __dirname + '/..'});
        });
        app.get('/getpreset/:id', (req, res) => {
            res.sendFile(req.params.id + '.json', {'root': __dirname + '/../presets'});
        });
        app.get('/ongoingdrafts', (req, res) => {
            res.json(draftsStore.getDraftIds());
        });

        app.use('/draft/[a-zA-Z]+', (req, res) => {
            res.sendFile(__dirname + '/index.html');
        });

        app.use(express.static('build'));
        app.use('/', (req, res) => {
            res.sendFile(__dirname + '/index.html');
        });

        io.on("connection", (socket: socketio.Socket) => {
            const draftIdRaw: string = socket.handshake.query.draftId;
            const draftId = Util.sanitizeDraftId(draftIdRaw);
            logger.info('Connection for draftId: %s', draftId, {draftId});

            const roomHost: string = `${draftId}-host`;
            const roomGuest: string = `${draftId}-guest`;
            const roomSpec: string = `${draftId}-spec`;

            if (!draftsStore.has(draftId)) {
                const path = `data/${draftId}.json`;
                if (fs.existsSync(path)) {
                    logger.info("Found recorded draft. Sending replay.", {draftId});
                    socket.emit('replay', JSON.parse(fs.readFileSync(path).toString('utf8')));
                } else {
                    logger.info("No recorded draft found.", {draftId});
                    socket.emit('message', 'This draft does not exist.');
                }
                logger.info("Disconnecting.", {draftId});
                socket.disconnect(true);
                return;
            }

            const rooms = Object.keys(socket.rooms);
            logger.debug("rooms: %s", JSON.stringify(rooms), {draftId});
            let yourPlayerType = Player.NONE;
            if (rooms.includes(roomHost)) {
                socket.join(roomHost); // async
                yourPlayerType = Player.HOST;
            } else if (rooms.includes(roomGuest)) {
                socket.join(roomGuest); // async
                yourPlayerType = Player.GUEST;
            } else {
                socket.join(roomSpec); // async
            }

            socket.on("set_role", (message: ISetRoleMessage, fn: (dc: IDraftConfig) => void) => {
                logger.info("Player wants to set own role: %s", JSON.stringify(message), {draftId});
                const role: Player = Util.sanitizeRole(message.role);
                let assignedRole = getAssignedRole(socket, roomHost, roomGuest);
                const rooms = Object.keys(socket.rooms);
                if (rooms.includes(roomSpec)) {
                    if (role === Player.HOST && !draftsStore.isPlayerConnected(draftId, role)) {
                        logger.info("Setting player role to 'HOST': %s", message.name, {draftId});
                        socket.join(roomHost); // async
                        socket.leave(roomSpec); // async
                        connectPlayer(draftId, Player.HOST, message.name);
                        assignedRole = Player.HOST;
                    } else if (role === Player.GUEST && !draftsStore.isPlayerConnected(draftId, role)) {
                        logger.info("Setting player role to 'GUEST': %s", message.name, {draftId});
                        socket.join(roomGuest); // async
                        socket.leave(roomSpec); // async
                        connectPlayer(draftId, Player.GUEST, message.name);
                        assignedRole = Player.GUEST;
                    } else {
                        logger.info("Setting role not possible: %s", role, {draftId});
                    }
                } else {
                    logger.info("Player is not registered as spectator currently. No action taken.", {draftId});
                }

                socket.nsp
                    .in(roomHost)
                    .in(roomGuest)
                    .in(roomSpec)
                    .emit("player_set_role", {name: message.name, playerType: assignedRole});
                fn({
                    ...draftsStore.getDraftViewsOrThrow(draftId).getDraftForPlayer(assignedRole),
                    yourPlayerType: assignedRole,
                });
            });

            socket.on("set_name", (message: ISetNameMessage, fn: () => void) => {
                logger.info("Player wants to set own name: %s", JSON.stringify(message), {draftId});
                let assignedRole = getAssignedRole(socket, roomHost, roomGuest);
                if (assignedRole === Player.HOST) {
                    logger.info("Setting HOST player name to: %s", message.name, {draftId});
                    setPlayerName(draftId, Player.HOST, message.name);
                } else if (assignedRole === Player.GUEST) {
                    logger.info("Setting GUEST player name to: %s", message.name, {draftId});
                    setPlayerName(draftId, Player.GUEST, message.name);
                } else {
                    logger.info("Player is not registered as HOST or GUEST currently. No action taken.", {draftId});
                    return;
                }

                socket.nsp
                    .in(roomHost)
                    .in(roomGuest)
                    .in(roomSpec)
                    .emit("player_set_name", {name: message.name, playerType: assignedRole});
                fn();
            });

            socket.on("ready", (message: {}, fn: (dc: IDraftConfig) => void) => {
                logger.info("Player indicates they are ready: %s", JSON.stringify(message), {draftId});
                let assignedRole: Player = Player.NONE;
                if (Object.keys(socket.rooms).includes(roomHost)) {
                    draftsStore.setPlayerReady(draftId, Player.HOST);
                    assignedRole = Player.HOST;
                } else if (Object.keys(socket.rooms).includes(roomGuest)) {
                    draftsStore.setPlayerReady(draftId, Player.GUEST);
                    assignedRole = Player.GUEST;
                }
                if (draftsStore.playersAreReady(draftId)) {
                    logger.info("Both Players are ready, starting countdown.", {draftId});
                    draftsStore.startCountdown(draftId, socket);
                    draftsStore.setStartTimestamp(draftId);
                }
                socket.nsp
                    .in(roomHost)
                    .in(roomGuest)
                    .in(roomSpec)
                    .emit("player_ready", {playerType: assignedRole});
                fn({
                    ...draftsStore.getDraftViewsOrThrow(draftId).getDraftForPlayer(assignedRole),
                    yourPlayerType: assignedRole
                });
            });

            socket.on("act", Listeners.actListener(draftsStore, draftId, validateAndApply, socket, roomHost, roomGuest, roomSpec));

            socket.on('disconnect', function () {
                const assignedRole = getAssignedRole(socket, roomHost, roomGuest);
                logger.info("Player disconnected: %s", assignedRole, {draftId});
                draftsStore.disconnectPlayer(draftId, assignedRole);
            });

            let draftState = {
                ...draftsStore.getDraftViewsOrThrow(draftId).getDraftForPlayer(Player.NONE),
                yourPlayerType: yourPlayerType
            };
            logger.info("Emitting draft_state: %s", JSON.stringify(draftState), {draftId});
            socket.emit("draft_state", draftState);
        });

        const httpServer = server.listen(port, () => {
            logger.info("Listening on: %d", port);
        });
        const httpServerAddr = httpServer.address();


        function validateAndApply(draftId: string, message: DraftEvent): ValidationId[] {
            return validator.validateAndApply(draftId, message);
        }

        setInterval(() => {
            draftsStore.purgeStaleDrafts()
        }, TWELVE_HOURS);

        return {httpServer, httpServerAddr, io};
    }
};