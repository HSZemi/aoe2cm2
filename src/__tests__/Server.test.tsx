import {default as io} from "socket.io-client"
import {DraftServer} from "../DraftServer";
import request from "request"
import Player from "../constants/Player";
import {IDraftConfig} from "../types/IDraftConfig";
import getPort from "get-port";
import Preset from "../models/Preset";
import {Barrier} from "../test/Barrier";

let hostSocket: any;
let clientSocket: any;
let spectatorSocket: any;
let httpServer: any;
let httpServerAddr: any;
let ioServer: any;
let draftId: string;

beforeAll((done) => {
    getPort().then((port: number) => {
        console.log("Got port: " + port);
        const serve = DraftServer.serve(port);
        httpServer = serve.httpServer;
        httpServerAddr = serve.httpServerAddr;
        ioServer = serve.io;
        done();
    });
});

afterAll((done) => {
    ioServer.close();
    httpServer.close();
    done();
});

function connect() {
    return io.connect(`http://[${httpServerAddr.address}]:${httpServerAddr.port}`, {
        query: {draftId: draftId},
        reconnectionDelay: 0,
        forceNew: true,
        transports: ['websocket'],
    });
}

beforeEach((done) => {
    const barrier = new Barrier(3, done);
    request.post(`http://[${httpServerAddr.address}]:${httpServerAddr.port}/draft/new`,
        {body: JSON.stringify({preset: Preset.SIMPLE}), headers: {'Content-Type': 'application/json; charset=UTF-8'}},
        (error, response, body) => {
            const draftIdContainer: { draftId: string } = JSON.parse(body);
            draftId = draftIdContainer.draftId;

            hostSocket = connect();
            hostSocket.on('connect', () => {
                barrier.trigger();
            });

            clientSocket = connect();
            clientSocket.on('connect', () => {
                barrier.trigger();
            });

            spectatorSocket = connect();
            spectatorSocket.on('connect', () => {
                barrier.trigger();
            });
        });
});

afterEach((done) => {
    if (hostSocket.connected) {
        hostSocket.disconnect();
    }
    if (clientSocket.connected) {
        clientSocket.disconnect();
    }
    if (spectatorSocket.connected) {
        spectatorSocket.disconnect();
    }
    done();
});

it('successful join gets a draft config', (done) => {
    hostSocket.emit('set_role', {name: 'Saladin', role: Player.HOST}, (data: IDraftConfig) => {
        data.startTimestamp = 0;
        expect(data).toMatchSnapshot();
        done();
    });
});


it('should send player_set_role when player sets role', (done) => {
    const barrier = new Barrier(2, done);
    hostSocket.once("player_set_role", (message: any) => {
        expect(message.name).toBe('Saladin');
        expect(message.playerType).toBe(Player.HOST);

        hostSocket.once("player_set_role", (message: any) => {
            expect(message.name).toBe('Barbarossa');
            expect(message.playerType).toBe(Player.GUEST);
            barrier.trigger();
        });
        clientSocket.emit('set_role', {name: 'Barbarossa', role: Player.GUEST}, () => {
        });
    });

    clientSocket.once("player_set_role", (message: any) => {
        expect(message.name).toBe('Saladin');
        expect(message.playerType).toBe(Player.HOST);

        clientSocket.once("player_set_role", (message: any) => {
            expect(message.name).toBe('Barbarossa');
            expect(message.playerType).toBe(Player.GUEST);
            barrier.trigger();
        });
    });

    hostSocket.emit('set_role', {name: 'Saladin', role: Player.HOST}, () => {
    });
});

it('players can change their names', (done) => {
    const barrier = new Barrier(4, done);
    hostSocket.once('player_set_name', (message: any) => {
        expect(message.name).toEqual('Attila The Hun');
        hostSocket.once('player_set_name', (message: any) => {
            expect(message.name).toEqual('Bleda The Hun');
            barrier.trigger();
        });
    });
    clientSocket.once('player_set_name', (message: any) => {
        expect(message.name).toEqual('Attila The Hun');
        clientSocket.once('player_set_name', (message: any) => {
            expect(message.name).toEqual('Bleda The Hun');
            barrier.trigger();
        });
    });
    spectatorSocket.once('player_set_name', (message: any) => {
        expect(message.name).toEqual('Attila The Hun');
        spectatorSocket.once('player_set_name', (message: any) => {
            expect(message.name).toEqual('Bleda The Hun');
            barrier.trigger();
        });
    });
    const newHostName = "Attila The Hun";
    const newGuestName = "Bleda The Hun";
    hostSocket.emit('set_role', {name: 'Saladin', role: Player.HOST}, () => {
        clientSocket.emit('set_role', {name: 'Barbarossa', role: Player.GUEST}, () => {
            clientSocket.emit('ready', {}, () => {
                hostSocket.emit('set_name', {
                    "name": newHostName
                }, () => {
                    hostSocket.emit('ready', {}, () => {
                        clientSocket.emit('set_name', {
                            "name": newGuestName
                        }, () => {
                            // done
                            const secondSpectatorSocket = io.connect(`http://[${httpServerAddr.address}]:${httpServerAddr.port}`, {
                                query: {draftId: draftId},
                                reconnectionDelay: 0,
                                forceNew: true,
                                transports: ['websocket'],
                            });
                            secondSpectatorSocket.once('draft_state', (message: any) => {
                                expect(message.nameHost).toEqual(newHostName);
                                expect(message.nameGuest).toEqual(newGuestName);
                                barrier.trigger();
                            });
                        });
                    });
                });
            });
        });
    });
});

it('fully execute sample draft', (done) => {
    const barrier = new Barrier(2, done);
    hostSocket.once('disconnect', () => {
        barrier.trigger();
    });
    clientSocket.once('disconnect', () => {
        barrier.trigger();
    });
    hostSocket.emit('set_role', {name: 'Saladin', role: Player.HOST}, () => {
        clientSocket.emit('set_role', {name: 'Barbarossa', role: Player.GUEST}, () => {
            clientSocket.emit('ready', {}, () => {
                hostSocket.emit('ready', {}, () => {
                    hostSocket.emit('act', {
                        "player": "HOST",
                        "actionType": "ban",
                        "civilisation": {"name": "Celts", "gameVersion": 1, "isRandomlyChosenCiv": false}
                    }, () => {
                        clientSocket.emit('act', {
                            "player": "GUEST",
                            "actionType": "ban",
                            "civilisation": {"name": "Celts", "gameVersion": 1, "isRandomlyChosenCiv": false}
                        }, () => {
                            clientSocket.emit('act', {
                                "player": "GUEST",
                                "actionType": "pick",
                                "civilisation": {"name": "Slavs", "gameVersion": 3, "isRandomlyChosenCiv": false}
                            }, () => {
                                hostSocket.emit('act', {
                                    "player": "HOST",
                                    "actionType": "pick",
                                    "civilisation": {"name": "Slavs", "gameVersion": 3, "isRandomlyChosenCiv": false}
                                }, () => {
                                    // preset done
                                });
                            });
                        });
                    });
                });
            });
        });
    });
});