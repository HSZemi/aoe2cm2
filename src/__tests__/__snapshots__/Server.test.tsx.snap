// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`successful join gets a draft config 1`] = `
Object {
  "events": Array [],
  "guestConnected": false,
  "guestReady": false,
  "hostConnected": true,
  "hostReady": false,
  "nameGuest": "â€¦",
  "nameHost": "Saladin",
  "nextAction": 0,
  "preset": Object {
    "encodedCivilisations": "7ffffffff",
    "name": "Simple Preset",
    "turns": Array [
      Object {
        "action": "BAN",
        "exclusivity": "NONEXCLUSIVE",
        "hidden": false,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "HOST",
      },
      Object {
        "action": "BAN",
        "exclusivity": "NONEXCLUSIVE",
        "hidden": false,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "GUEST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "NONEXCLUSIVE",
        "hidden": false,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "GUEST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "NONEXCLUSIVE",
        "hidden": false,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "HOST",
      },
    ],
  },
  "startTimestamp": 0,
  "yourPlayerType": "HOST",
}
`;
